proposals/list-patterns.md

```diff
@@ -0,0 +1,122 @@
+# List patterns
+
+## Summary
+
+Lets you to match an array or a list with a sequence of patterns e.g. `array is {1, 2, 3}` will match an integer array of the length three with 1, 2, 3 as its elements, respectively.
+
+## Detailed Design
+
+The pattern syntax is modified as follow:
+
+```antlr
+recursive_pattern
+  : type? positional_pattern_clause? length_pattern_clause? property_or_list_pattern_clause? simple_designation?
+  ;
+
+length_pattern_clause
+  : '[' pattern ']'
+  ;
+
+property_or_list_pattern_clause
+  : list_pattern_clause
+  | property_pattern_clause
+  ;
+
+property_pattern_clause
+  : '{' (subpattern (',' subpattern)* ','?)? '}'
+  ;
+
+list_pattern_clause
+  : '{' pattern (',' pattern)* ','? '}'
+  ;
+
+slice_pattern
+  : '..' negated_pattern?
+  ;
+
+primary_pattern
```

Comment:
What's the benefit of overloading `recursive_pattern`?
What if we instead `primary_pattern` to have a choice of `list_pattern` defined as  `length_pattern_clause? list_pattern_clause`?

The `positional_pattern_clause` and `simple_designation` portions would not be allowed in a list pattern anyways, right?

Reply:
>The positional_pattern_clause and simple_designation portions would not be allowed in a list pattern anyways, right?

Actually that's the case. LDT wants this to be a part of recursive patterns mainly to allow a type e.g. `IEnumerable {4}`, but positionals and designations would be allowed as well. Same for length patterns. The only limitation is that list and property clauses cannot appear side by side, otherwise all combinations are valid. 

Note that `[1] {}` would NOT be an error because the length doesn't match, rather `{}` is still (and always) a property pattern clause. Fred suggested we warn here to be not confused that way.

Reply:
Ah, thanks. Could we include an example with the type and other parts?
I must say that re-using the recursive pattern still seems strange to me. Why allow combining a positional check with a list check, but not a pattern check with a list check.

By the way, I think the patterns spec was re-writing without `recursive_pattern` as it had all its parts optional. See [here](https://github.com/dotnet/csharplang/blob/a3d67d141501101af9dbb9e8192ca8b4c717e316/proposals/csharp-8.0/patterns.md#patterns) for way the patterns were divided.

I'd propose we do something similar for length and list patterns, which disallows their combination with positional patterns.